File: A08 Dynamic Programming/2035. Partition Array Into Two Arrays to Minimize Sum Difference/DP solution.java
/*
Time Complexity:
Space Complexity:
DP solution, but test case didn't work for negative number [-36, 36]
https://www.youtube.com/watch?v=FB0KUhsxXGY
*/ 

class Solution {
    public int minimumDifference(int[] nums) {
        int n = nums.length; 
        int sum = 0; 
        
        for (int num : nums){
            sum += num;
        }
        
        //fill the dynamic table
        boolean[][] dp = new boolean [n+1][sum +1];  
        //boolean[][] dp = new boolean [n+1][sum/2 +1]; //work better?
        for(int i = 0; i <= n ; i++){
            for (int j = 0; j <= sum; j++){
                //for (int j = 0; j <= sum/2; j++){
                if (j == 0){
                    dp[i][j] = true;
                }else if(i == 0){
                    dp[i][j] = false;
                } else if(nums[i -1] > j){
                    dp[i][j] = dp[i-1][j]; 
                } else{
                    dp[i][j] = dp[i-1][j] || dp[i-1][j- nums[i -1]];
                }
            }
        }
        
        //find min difference value 
        int diff = Integer.MAX_VALUE;
        for(int i =0; i <= sum/2; i++){
            int first = i; 
            int second = sum - i; 
            
            if (dp[n][i] == true  &&diff > Math.abs(first - second) ){
               // if (first > 0){
                diff = Math.abs(first - second);
               // } else{
                 //   diff = Math.abs(second - first);
                //}
            }
        }
        return diff; 
    }
}

File: A08 Dynamic Programming/198. House Robber/DP solution.py
###########
Time Complexity:
Space Complexity:


#############
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        if length==0:
            return 0
        if length==1:
            return nums[0]
        if length==2:
            return max(nums)
			
        dp = [0]*length # assign dp array
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])
        
        for i in range(2, length):
            dp[i] = max(dp[i-2]+nums[i], dp[i-1])
        print(dp)
        
        return max(dp)


#########
class Solution:
    def rob(self, nums: List[int]) -> int:
        if (len(nums) == 0): 
            return 0
        if (len(nums) == 1):
            return nums[0]
        if (len(nums) == 2):
            return max(nums[0], nums[1])  #return max(nums)
            
        dp = [0 for i in range (len(nums))]  #dp = [0]*length # assign dp array
        for i in range(len(nums)):
            dp[0] = nums[0]
            dp[1] = max(nums[0], nums[1])
            dp[i] = max(nums[i] + dp[i-2] , dp[i-1])
        return dp[len(nums) - 1] 